Frequent trickster crashes during prometheus infrastructure patching (invalid memory address or nil pointer dereference)

We are running trickster (v2.0.0-beta2) in front of several different prometheus running on OpenShift clusters and in front of different loadbalancer. Every time we patch one of the OpenShift environment and their loadbalancers, trickster crashes frequently, between every few minutes and every 30 minutes. We are using Grafana alerting which is constantly querying trickster so patching one environment can cause alerts for all others behind the same trickster instance depending on the timing.

With debug logging and tracing enabled, we can see the following in the log:

```
time=2023-01-24T21:42:03.072489067Z app=trickster level=debug event="delta routine completed" cacheKey=REDACTED_URL2:4500.dpc.d173064bcdaa3f6cc153817b1567991f cacheStatus=phit reqStart=1674596460 reqEnd=1674596520 extentsFetched=1674596461000-1674596520000 caller=proxy/engines/deltaproxycache.go:512
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"920985e77c4cea72","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"ac4c973845869122","SpanKind":1,"Name":"QueryCache","StartTime":"2023-01-24T21:42:03.080497122Z","EndTime":"2023-01-24T21:42:03.080529635Z","Attributes":[{"Key":"cache.status","Value":{"Type":"STRING","Value":"hit"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.080545157Z app=trickster level=debug event="memory cache retrieve" cacheKey=REDACTED_URL3:4500.dpc.6c1487c286555887d5e426a0cec9e471 caller=cache/memory/memory.go:148
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"d0251dfdeaf0e160","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"35d8f8be91c4de59","SpanKind":1,"Name":"ProxyRequest","StartTime":"2023-01-24T21:42:03.08072198Z","EndTime":"2023-01-24T21:42:03.093722847Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"35d8f8be91c4de59","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"dfb89f7f8187c400","SpanKind":1,"Name":"PrepareFetchReader","StartTime":"2023-01-24T21:42:03.080694073Z","EndTime":"2023-01-24T21:42:03.093871854Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"dfb89f7f8187c400","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"ac4c973845869122","SpanKind":1,"Name":"FetchRange","StartTime":"2023-01-24T21:42:03.080688907Z","EndTime":"2023-01-24T21:42:03.093993957Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.094109614Z app=trickster level=debug event="upstream request" handlerName=query_range userAgent=Grafana/9.3.2 caller=proxy/engines/access_logs.go:27 size=64 durationMS=13 backendName=REDACTED_DOMAIN3 backendProvider=prometheus method=GET uri="https://REDACTED_URL3:4500/api/v1/query_range?end=1674596520&query=sum%28node_load15%29+by+%28instance%29+%2F+count%28node_cpu_info%29+by+%28instance%29+%3E+8&start=1674596461&step=1" code=200
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"c09e8d2d8bceb4d8","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"ac4c973845869122","SpanKind":1,"Name":"WriteCache","StartTime":"2023-01-24T21:42:03.094169803Z","EndTime":"2023-01-24T21:42:03.09419279Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.094222145Z app=trickster level=debug event="memorycache cache store" cacheName=default cacheKey=REDACTED_URL3:4500.dpc.6c1487c286555887d5e426a0cec9e471 length=0 ttl=6h0m0s is_direct=true caller=cache/memory/memory.go:99
time=2023-01-24T21:42:03.094328218Z app=trickster level=debug event="delta routine completed" cacheStatus=phit reqStart=1674596460 reqEnd=1674596520 extentsFetched=1674596461000-1674596520000 caller=proxy/engines/deltaproxycache.go:512 cacheKey=REDACTED_URL3:4500.dpc.6c1487c286555887d5e426a0cec9e471
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"ac4c973845869122","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"06e84eda86dee4f8","SpanKind":1,"Name":"DeltaProxyCacheRequest","StartTime":"2023-01-24T21:42:03.080404116Z","EndTime":"2023-01-24T21:42:03.095327304Z","Attributes":[{"Key":"cache.status","Value":{"Type":"STRING","Value":"phit"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":3,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"a8e852a98dd3331680c9b82bc1e7880c","SpanID":"06e84eda86dee4f8","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"0000000000000000","SpanKind":1,"Name":"request","StartTime":"2023-01-24T21:42:03.080375987Z","EndTime":"2023-01-24T21:42:03.095440469Z","Attributes":[{"Key":"http.method","Value":{"Type":"STRING","Value":"GET"}},{"Key":"http.target","Value":{"Type":"STRING","Value":"/REDACTED_DOMAIN3/api/v1/query_range?end=1674596520\u0026query=sum%28node_load15%29+by+%28instance%29+%2F+count%28node_cpu_info%29+by+%28instance%29+%3E+8\u0026start=1674596460\u0026step=1"}},{"Key":"http.user_agent","Value":{"Type":"STRING","Value":"Grafana/9.3.2"}},{"Key":"http.scheme","Value":{"Type":"STRING","Value":"https"}},{"Key":"http.host","Value":{"Type":"STRING","Value":"dcsgwlgpmon02.dcsg.govt.container:8483"}},{"Key":"http.flavor","Value":{"Type":"STRING","Value":"1.1"}},{"Key":"net.transport","Value":{"Type":"STRING","Value":"IP.TCP"}},{"Key":"net.peer.ip","Value":{"Type":"STRING","Value":"10.0.2.100"}},{"Key":"net.peer.port","Value":{"Type":"INT64","Value":54024}},{"Key":"net.host.name","Value":{"Type":"STRING","Value":"dcsgwlgpmon02.dcsg.govt.container"}},{"Key":"net.host.port","Value":{"Type":"INT64","Value":8483}},{"Key":"backend.name","Value":{"Type":"STRING","Value":"REDACTED_DOMAIN3"}},{"Key":"backend.provider","Value":{"Type":"STRING","Value":"prometheus"}},{"Key":"router.path","Value":{"Type":"STRING","Value":"/api/v1/query_range"}},{"Key":"cache.name","Value":{"Type":"STRING","Value":"default"}},{"Key":"cache.provider","Value":{"Type":"STRING","Value":"memory"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"55cae012a99b6f4a","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"9f7c7c6d08370fff","SpanKind":1,"Name":"QueryCache","StartTime":"2023-01-24T21:42:03.220716987Z","EndTime":"2023-01-24T21:42:03.220765616Z","Attributes":[{"Key":"cache.status","Value":{"Type":"STRING","Value":"hit"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.220789157Z app=trickster level=debug event="memory cache retrieve" cacheKey=REDACTED_URL2:4500.dpc.1b30325cc4c4b9996a2b0590355cf12e caller=cache/memory/memory.go:148
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"c807ceb9245536f7","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"c5e2021ada172fd7","SpanKind":1,"Name":"ProxyRequest","StartTime":"2023-01-24T21:42:03.221016141Z","EndTime":"2023-01-24T21:42:03.272770212Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"c5e2021ada172fd7","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"3e4f3f3080eab566","SpanKind":1,"Name":"PrepareFetchReader","StartTime":"2023-01-24T21:42:03.220986555Z","EndTime":"2023-01-24T21:42:03.27294653Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"3e4f3f3080eab566","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"9f7c7c6d08370fff","SpanKind":1,"Name":"FetchRange","StartTime":"2023-01-24T21:42:03.220980081Z","EndTime":"2023-01-24T21:42:03.27308766Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.273273427Z app=trickster level=debug event="upstream request" backendName=REDACTED_DOMAIN2 uri="https://REDACTED_URL2:4500/api/v1/query_range?end=1674596520&query=kube_node_status_condition%7Bcondition%3D%22Ready%22%2Cstatus%3D%22true%22%7D+%3D%3D+0+OR+absent%28kube_node_status_condition%7Bcondition%3D%22Ready%22%2Cstatus%3D%22true%22%7D%29&start=1674596461&step=1" userAgent=Grafana/9.3.2 durationMS=52 size=64 caller=proxy/engines/access_logs.go:27 backendProvider=prometheus handlerName=query_range method=GET code=200
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"9f7c7c6d08370fff","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"7d7af099844001db","SpanKind":1,"Name":"DeltaProxyCacheRequest","StartTime":"2023-01-24T21:42:03.220607776Z","EndTime":"2023-01-24T21:42:03.273477666Z","Attributes":[{"Key":"cache.status","Value":{"Type":"STRING","Value":"phit"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":3,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"24cd55f1bba5f77e","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"9f7c7c6d08370fff","SpanKind":1,"Name":"WriteCache","StartTime":"2023-01-24T21:42:03.273423808Z","EndTime":"2023-01-24T21:42:03.273480051Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.273559957Z app=trickster level=debug event="memorycache cache store" ttl=6h0m0s is_direct=true caller=cache/memory/memory.go:99 cacheName=default cacheKey=REDACTED_URL2:4500.dpc.1b30325cc4c4b9996a2b0590355cf12e length=0
[{"SpanContext":{"TraceID":"9429ac72a6fac6a29af7f34da0ed243c","SpanID":"7d7af099844001db","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"0000000000000000","SpanKind":1,"Name":"request","StartTime":"2023-01-24T21:42:03.22057606Z","EndTime":"2023-01-24T21:42:03.273539121Z","Attributes":[{"Key":"http.method","Value":{"Type":"STRING","Value":"GET"}},{"Key":"http.target","Value":{"Type":"STRING","Value":"/REDACTED_DOMAIN2/api/v1/query_range?end=1674596520\u0026query=kube_node_status_condition%7Bcondition%3D%22Ready%22%2Cstatus%3D%22true%22%7D+%3D%3D+0+OR+absent%28kube_node_status_condition%7Bcondition%3D%22Ready%22%2Cstatus%3D%22true%22%7D%29\u0026start=1674596460\u0026step=1"}},{"Key":"http.user_agent","Value":{"Type":"STRING","Value":"Grafana/9.3.2"}},{"Key":"http.scheme","Value":{"Type":"STRING","Value":"https"}},{"Key":"http.host","Value":{"Type":"STRING","Value":"dcsgwlgpmon02.dcsg.govt.container:8483"}},{"Key":"http.flavor","Value":{"Type":"STRING","Value":"1.1"}},{"Key":"net.transport","Value":{"Type":"STRING","Value":"IP.TCP"}},{"Key":"net.peer.ip","Value":{"Type":"STRING","Value":"10.0.2.100"}},{"Key":"net.peer.port","Value":{"Type":"INT64","Value":45430}},{"Key":"net.host.name","Value":{"Type":"STRING","Value":"dcsgwlgpmon02.dcsg.govt.container"}},{"Key":"net.host.port","Value":{"Type":"INT64","Value":8483}},{"Key":"backend.name","Value":{"Type":"STRING","Value":"REDACTED_DOMAIN2"}},{"Key":"backend.provider","Value":{"Type":"STRING","Value":"prometheus"}},{"Key":"router.path","Value":{"Type":"STRING","Value":"/api/v1/query_range"}},{"Key":"cache.name","Value":{"Type":"STRING","Value":"default"}},{"Key":"cache.provider","Value":{"Type":"STRING","Value":"memory"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.273730117Z app=trickster level=debug event="delta routine completed" cacheKey=REDACTED_URL2:4500.dpc.1b30325cc4c4b9996a2b0590355cf12e cacheStatus=phit reqStart=1674596460 reqEnd=1674596520 extentsFetched=1674596461000-1674596520000 caller=proxy/engines/deltaproxycache.go:512
[{"SpanContext":{"TraceID":"aa7c911b8f2ae4999204bc7a49288bc8","SpanID":"4216b8b9d050e7bd","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"b0423597893a38d5","SpanKind":1,"Name":"QueryCache","StartTime":"2023-01-24T21:42:03.289780039Z","EndTime":"2023-01-24T21:42:03.289814613Z","Attributes":[{"Key":"cache.status","Value":{"Type":"STRING","Value":"hit"}}],"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.289870038Z app=trickster level=debug event="memory cache retrieve" caller=cache/memory/memory.go:148 cacheKey=REDACTED_URL1:4500.dpc.558e54d875c7bb36cb59f54b1984a086
[{"SpanContext":{"TraceID":"aa7c911b8f2ae4999204bc7a49288bc8","SpanID":"bf4eb31c78d3449b","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"f34a040fca8a2eb5","SpanKind":1,"Name":"ProxyRequest","StartTime":"2023-01-24T21:42:03.290044492Z","EndTime":"2023-01-24T21:42:03.298783509Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":0,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"aa7c911b8f2ae4999204bc7a49288bc8","SpanID":"f34a040fca8a2eb5","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"ea67c6feb9556ebe","SpanKind":1,"Name":"PrepareFetchReader","StartTime":"2023-01-24T21:42:03.290004434Z","EndTime":"2023-01-24T21:42:03.298949902Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
[{"SpanContext":{"TraceID":"aa7c911b8f2ae4999204bc7a49288bc8","SpanID":"ea67c6feb9556ebe","TraceFlags":1,"TraceState":null,"Remote":false},"ParentSpanID":"b0423597893a38d5","SpanKind":1,"Name":"FetchRange","StartTime":"2023-01-24T21:42:03.289998024Z","EndTime":"2023-01-24T21:42:03.299085473Z","Attributes":null,"MessageEvents":null,"Links":null,"StatusCode":"Unset","StatusMessage":"","HasRemoteParent":false,"DroppedAttributeCount":0,"DroppedMessageEventCount":0,"DroppedLinkCount":0,"ChildSpanCount":1,"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"trickster"}}],"InstrumentationLibrary":{"Name":"default","Version":""}}]
time=2023-01-24T21:42:03.299289145Z app=trickster level=debug event="upstream request" backendName=REDACTED_DOMAIN1 backendProvider=prometheus handlerName=query_range size=64 durationMS=9 caller=proxy/engines/access_logs.go:27 method=GET uri="https://REDACTED_URL1:4500/api/v1/query_range?end=1674596520&query=kube_pod_container_status_ready%7Bpod%3D~%22trident-csi-.%2A%22%7D+%3D%3D+0+OR+absent%28kube_pod_container_status_ready%7Bpod%3D~%22trident-csi-.%2A%22%7D%29&start=1674596461&step=1" userAgent=Grafana/9.3.2 code=200
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xb1d841]

goroutine 1496 [running]:
github.com/tricksterproxy/trickster/pkg/timeseries/dataset.(*DataSet).InjectTags.func1(0xc000430720, 0xc00007fcb0, 0xc000430710, 0x0)
        /home/runner/work/trickster/trickster/pkg/timeseries/dataset/tags.go:41 +0x41
created by github.com/tricksterproxy/trickster/pkg/timeseries/dataset.(*DataSet).InjectTags
        /home/runner/work/trickster/trickster/pkg/timeseries/dataset/tags.go:39 +0xde

```

--- a/pkg/timeseries/dataset/dataset.go
+++ b/pkg/timeseries/dataset/dataset.go
@@ -456,6 +456,9 @@ func (ds *DataSet) DefaultRangeCropper(e timeseries.Extent) {
 func (ds *DataSet) SeriesCount() int {
 	var cnt int
 	for i := range ds.Results {
+		if ds.Results[i] == nil {
+			continue
+		}
 		cnt += len(ds.Results[i].SeriesList)
 	}
 	return cnt

--- a/pkg/timeseries/dataset/dataset.go
+++ b/pkg/timeseries/dataset/dataset.go
@@ -99,6 +99,9 @@ func (ds *DataSet) CroppedClone(e timeseries.Extent) timeseries.Timeseries {
 	// range, return empty set and bail
 	if ds.ExtentList.OutsideOf(e) {
 		for i := range ds.Results {
+			if ds.Results[i] == nil {
+				continue
+			}
 			clone.Results[i] = &Result{
 				StatementID: ds.Results[i].StatementID,
 				Error:       ds.Results[i].Error,
@@ -196,6 +199,9 @@ func (ds *DataSet) Clone() timeseries.Timeseries {
 	}
 
 	for i := range ds.Results {
+		if ds.Results[i] == nil {
+			continue
+		}
 		clone.Results[i] = ds.Results[i].Clone()
 	}
 	return clone
@@ -367,6 +373,9 @@ func (ds *DataSet) DefaultRangeCropper(e timeseries.Extent) {
 	// The DataSet has no extents, so no need to do anything
 	if x == 0 {
 		for i := range ds.Results {
+			if ds.Results[i] == nil {
+				continue
+			}
 			ds.Results[i].SeriesList = make([]*Series, 0)
 		}
 		ds.ExtentList = timeseries.ExtentList{}
@@ -376,6 +385,9 @@ func (ds *DataSet) DefaultRangeCropper(e timeseries.Extent) {
 	// range, return empty set and bail
 	if ds.ExtentList.OutsideOf(e) {
 		for i := range ds.Results {
+			if ds.Results[i] == nil {
+				continue
+			}
 			ds.Results[i].SeriesList = make([]*Series, 0)
 		}
 		ds.ExtentList = timeseries.ExtentList{}
@@ -401,6 +413,9 @@ func (ds *DataSet) DefaultRangeCropper(e timeseries.Extent) {
 	endNS := epoch.Epoch(e.End.UnixNano())
 
 	for i := range ds.Results {
+		if ds.Results[i] == nil {
+			continue
+		}
 		var wg sync.WaitGroup
 		if len(ds.Results[i].SeriesList) == 0 {
 			continue
@@ -450,6 +465,9 @@ func (ds *DataSet) SeriesCount() int {
 func (ds *DataSet) ValueCount() int64 {
 	var cnt int64
 	for i := range ds.Results {
+		if ds.Results[i] == nil {
+			continue
+		}
 		if len(ds.Results[i].SeriesList) == 0 {
 			continue
 		}

--- a/pkg/timeseries/dataset/dataset_test.go
+++ b/pkg/timeseries/dataset/dataset_test.go
@@ -99,6 +99,7 @@ func testDataSet2() *DataSet {
 		StatementID: 0,
 		SeriesList: []*Series{
 			{sh1, newPoints(), s},
+			nil,
 		},
 	}
 
@@ -113,7 +114,7 @@ func testDataSet2() *DataSet {
 
 	ds := &DataSet{
 		TimeRangeQuery: &timeseries.TimeRangeQuery{Step: time.Duration(5 * timeseries.Second)},
-		Results:        []*Result{r1, r2},
+		Results:        []*Result{r1, r2, nil},
 		ExtentList:     timeseries.ExtentList{timeseries.Extent{Start: time.Unix(5, 0), End: time.Unix(30, 0)}},
 	}
 

--- a/pkg/timeseries/dataset/dataset_test.go
+++ b/pkg/timeseries/dataset/dataset_test.go
@@ -214,7 +214,7 @@ func TestMerge(t *testing.T) {
 
 	ds.Merge(false, ds2)
 
-	if ds.SeriesCount() != 4 {
+	if ds.SeriesCount() != 5 {
 		t.Errorf("expected %d got %d", 4, ds.SeriesCount())
 	}
 }

--- a/pkg/timeseries/dataset/series_list.go
+++ b/pkg/timeseries/dataset/series_list.go
@@ -75,6 +75,9 @@ func (sl SeriesList) Equal(sl2 SeriesList) bool {
 		return false
 	}
 	for i, v := range sl {
+		if v == nil {
+			continue
+		}
 		if v.Header.CalculateHash() != sl2[i].Header.CalculateHash() {
 			return false
 		}

--- a/pkg/timeseries/dataset/tags.go
+++ b/pkg/timeseries/dataset/tags.go
@@ -34,7 +34,13 @@ func (ds *DataSet) InjectTags(tags Tags) {
 	var wg sync.WaitGroup
 	var mtx sync.Mutex
 	for _, r := range ds.Results {
+		if r == nil {
+			continue
+		}
 		for _, s := range r.SeriesList {
+			if s == nil {
+				continue
+			}
 			wg.Add(1)
 			go func(s1 *Series) {
 				mtx.Lock()

